harness/determined/pytorch/deepspeed/_deepspeed_context.py

```diff
@@ -0,0 +1,352 @@
+        pytorch._PyTorchReducerContext.__init__(self, self.distributed._zmq_allgather)
+
+        self._init_device()
+
+        # Track which types we have issued warnings for in to_device().
+        self._to_device_warned_types = set()  # type: Set[Type]
+
+        # DeepSpeed supports mixed precision through NVidia Apex AMP.  ZeRO optimizer requires
+        # Apex AMP and cannot be used with more complex AMP modes.
+        apex_available = importlib.util.find_spec("apex") is not None
+        if not apex_available:
+            logging.warning(
+                "Missing package APEX is required for ZeRO optimizer support through DeepSpeed."
+            )
+
+        # The following attributes are initialized during the lifetime of
+        # a DeepSpeedTrialContext.
+        self.models = []  # type: List[nn.Module]
+        self._epoch_len = None  # type: Optional[int]
+
+        self._loss_ids = {}  # type: Dict[torch.Tensor, int]
+        self._last_backward_batch_idx = None  # type: Optional[int]
+        self._current_batch_idx = None  # type: Optional[int]
+
+        self._determined_profiler = None  # type: Optional[profiler.ProfilerAgent]
+        self._mpu = det_ds.DeterminedModelParallelUnit(
+            self.distributed
+        )  # type: det_ds.ModelParallelUnit
+        self._called_wrap_mpu = False
+        # We will compute these as model engines and/or mpus are passed to the context.
+        self._global_batch_size = self.env.global_batch_size
+        self._per_slot_batch_size = self.env.per_slot_batch_size
+        self._train_micro_batch_size_per_gpu = None  # type: Optional[int]
+        self._num_micro_batches_per_slot = None  # type: Optional[int]
+        self._use_pipeline_parallel = False
+        self._data_repro_checks_disabled = False
+        self._manual_grad_accumulation = False
+
+        self._check_experiment_config()
+
+    def _check_experiment_config(self) -> None:
+        """
+        Check if the user specified options in optimizations that are incompatible with
+        DeepSpeedTrial.
+        """
+        optimizations_config = self.env.experiment_config.get_optimizations_config()
+        if not optimizations_config.get("average_training_metrics", False):
+            logging.warning(
+                "DeepSpeedTrial always averages training metrics across data parallel ranks."
+            )
+        self._average_training_metrics = True
+        mixed_precision_val = optimizations_config.get("mixed_precision", "O0")
+        if mixed_precision_val != "O0":
+            raise det.errors.InvalidExperimentException(
+                "Mixed precision is specified through the deepspeed config instead of the "
+                "Determined experiment config.",
+            )
+        aggregation_frequency = optimizations_config.get("aggregation_frequency", 1)
+        if aggregation_frequency > 1:
+            raise det.errors.InvalidExperimentException(
```

Comment:
Is this all of the things that we have in the experiment config which must be duplicated into the deepspeed config:
- global_batch_size
- mixed_precision
- average_training_metrics
- aggregation_frequency

Are there others?

Reply:
These are the ones that overlap.  The other fields in `optimizations` do not apply.  Do you think we should add a check for those?

Reply:
as long as we disallow duplicate configs (and get rid of the overall global_batch_size requirement) I think this approach is fine.